DA-327: CBImport Wizard

teetangh opened the pull request.

deniswsrosa reviewed (changes requested):
I can't start reviewing your code until you at least fixed the basic things:

- Remove commented code - IMO, good programmers don't ever push commented code unless there is a good reason for it

teetangh commented:
> I can't start reviewing your code until you have fixed the basics:
> 
> * Remove commented code - Don't ever push commented code unless there is a good reason for it
> * Fix formatting. A number of lines in your PR, nothing changed other than the formatting, if you are using VSCode, it is introducing a number of unnecessary changes. You need to adjust that or revert the formatting changes manually;
> * Separate UI from Business Logic - Have you ever heard of  Model-View-Controller (MVC)? Whenever your class has more than +700 lines, you should consider if what you are doing is correct. In your case, you are mixing the UI code with the backend code .  You need to move your cbexport logic to the CBExport class, and the UI should just call it.
> * We talked multiple times that CBImport must not create any indexes, yet your code has some logic to create indexes
> 
> Feels to me that you are not reviewing your own PR first.

thank you for your feedback.
I have removed commented code, fixed formatting as per intellij, separated the UI from the backend, removed index creation and also troubleshot advanced options in cbimport and enabled verbose logging

deniswsrosa commented on src/main/java/com/couchbase/intellij/tools/dialog/ImportDialogController.java:
```diff
@@ -0,0 +1,571 @@
+    protected static String targetScopeField;
+    protected static String targetCollectionField;
+
+    protected static void addListeners(ImportDialog importDialog) {
```
Updatesummary should called whenever a entity is updated. It should scan all values and show the result.

deniswsrosa reviewed (changes requested):
Refactoring is needed

deniswsrosa commented:
![Screenshot 2023-08-03 at 3 52 48 PM](https://github.com/couchbaselabs/couchbase_jetbrains_plugin/assets/216006/5e92fa92-a127-4ed3-b971-56f7b4d40455)

- There is no default choice select
- The mockup had an "?" icon, so the user could hover it and understand what that option means
- We discussed many times that the form should not be in the center, but in the top instead.  Because as you select "Dynamic scope and collection", the form is pushed up:
![Screenshot 2023-08-03 at 4 05 46 PM](https://github.com/couchbaselabs/couchbase_jetbrains_plugin/assets/216006/538edb00-ea1f-4f0f-9713-2e2f9960d910)

deniswsrosa commented:
We have the same problem on the other pages. Look how much the form moves depending on the option that I pick:
![Screenshot 2023-08-03 at 4 09 13 PM](https://github.com/couchbaselabs/couchbase_jetbrains_plugin/assets/216006/d565434e-9b37-410b-9614-cf03a5934d4b)
![Screenshot 2023-08-03 at 4 09 07 PM](https://github.com/couchbaselabs/couchbase_jetbrains_plugin/assets/216006/36261187-ad2d-4281-8f76-3622a539a2fd)

deniswsrosa commented:
Your code doesn't support - https://docs.couchbase.com/server/current/tools/cbimport-json.html#key-generators

![Screenshot 2023-08-03 at 4 16 30 PM](https://github.com/couchbaselabs/couchbase_jetbrains_plugin/assets/216006/d35d78b9-6093-441e-b915-e2c6f09a1455)

deniswsrosa commented:
![Screenshot 2023-08-03 at 4 19 46 PM](https://github.com/couchbaselabs/couchbase_jetbrains_plugin/assets/216006/3afbe804-ab0f-4e99-b5e1-4bfc3613ce59)
I can advance to the next step without specifying anything

deniswsrosa commented:
![Screenshot 2023-08-03 at 4 20 39 PM](https://github.com/couchbaselabs/couchbase_jetbrains_plugin/assets/216006/97e808d2-b29d-435b-8d62-4d97b085d177)
When the

teetangh commented:
Based on the feedback i received yesterday

1. [x] The UI is now clearly separated from the business logic.(no more ImportDialogcontroller the Api for that is in the CBImport as you explained. The Api just receives the command and executes it. Does need anything from the UI). All the UI stuff is in ImportDialog.java
2. [x] The Process Builder Command List has been made into an ArrayList as you described instead of a continuous chain of processBuilder.command().add(“something”) just like in pillow fight
3. [x] The createCenterPanel is no longer a bunch of components added together in a marathon. But instead has been clearly separated into 5 clean, simple functions createDatasetPanel, createTargetPanel, createKeyPanel, createAdvancedPanel, createSummaryPanel so that each function is simpler to read and debug
4. [x] Validation of each of the fields in each of the pages have been done thoroughly(Not only their respective data types but only also logically correct values)
5. [x] The Visible-Enable duo no longer exists and only visibility has been considered.
6. [x] The Fields are bordered with red in case a field is invalid and the reason for that has been described in an error label below
7. [x] The Next, Back and Cancel Buttons now work as expected with Next Button being greyed out when input is invalid/empty. The Back Button is invisible on the first page and Next Button changes to Import on the last page. Clever Browsing through Next and Back Button won’t allow the user to bypass field validation or vice versa ie deem the earlier validated data as invalid and block the next button
8. [x] The Action Events have now been converted to handle Events as discussed and all the event handlers have been extracted from the createCenterPanel UI and well organised in addListeners method.
9. [x] The Event listeners that were repetitive/common have been aggregated to produce one Event Listener
10. [x] The updateSummary() method has been changed to calculateSummary to prevent repetitive live-stream like updation of Summary as discussed.
11. [x] The EntityType enum has been removed from types and introduced in NewEntityCreationDialog since its a feature linked to that
12. [x] Dataset field is not only checked for json/csv file format. But also, if json file contents are invalid the parsing fails
13. [x] The number of documents in the json file are also counted
14. [x] The Help Icon for each Label has been introduced with appropriate hint messages
15. [x] All Sysouts have been removed
16. [x] The custom expression #UUID# and #MONO_INCR# will be simply passed onto cbimport command shell utility and the key generators will generate the relevant values in realtime. I cannot produce these values before hand thus left a placeholder in key preview area. Will add an info label so that the user doesn’t have the same worry
17. [x] Mockup: The key options was earlier to the left of the radio buttons. Now it’s been placed on the top.
18. [ ] Mockup: working on the moving the content to top of the page. Instead of center.

teetangh commented:
> ![Screenshot 2023-08-03 at 1 23 37 PM](https://user-images.githubusercontent.com/216006/258096503-f58726e3-5201-4acb-b6e6-93aa4596c8c3.png) The back button should not be visible on the first panel. @teetangh

validated and resolved

deniswsrosa reviewed (changes requested):
Unused Methods, Unused variables, Out-of-memory while trying to load a simple export, the IDE shows many warnings and no proper separation of concerns. Unfortunately, it is the only feature that won't be included in the 1.0.2

teetangh commented on src/main/java/com/couchbase/intellij/tools/dialog/ImportDialog.java:
```diff
@@ -0,0 +1,1554 @@
+            boolean verboseCheckSelected = verboseCheck.isSelected();
+
+            List<String> commandList = new ArrayList<>();
+            commandList.add(CBTools.getTool(CBTools.Type.CB_IMPORT).getPath());
```
Implemented Builder Pattern

teetangh commented on src/main/java/com/couchbase/intellij/tools/dialog/ImportDialog.java:
```diff
@@ -0,0 +1,1554 @@
+            } else if (keyGroupSelection == useFieldValueRadio.getModel()) {
+                commandList.add("--generate-key");
+                String fieldName = fieldNameText;
+                commandList.add(checkForDelimiters("%" + fieldName + "%", '#'));
```
removed the intermediary variable

teetangh commented on src/main/java/utils/FileUtils.java:
```diff
@@ -39,6 +53,142 @@ public static String readLastLine(String filePath) throws IOException {
         return lastLine;
     }
 
+    public static String readLine(String filePath, int lineNumber) throws IOException {
```
removed the extra methods and attributes

teetangh commented on src/main/java/utils/FileUtils.java:
```diff
@@ -39,6 +53,142 @@ public static String readLastLine(String filePath) throws IOException {
+        return isValidField;
+    }
+
+    public static String validateAndDetectCouchbaseJsonFormat(String filePath) throws IOException {
```
changed the naming

teetangh commented on src/main/java/utils/FileUtils.java:
```diff
@@ -39,6 +53,142 @@ public static String readLastLine(String filePath) throws IOException {
+            if (firstLine != null && firstLine.trim().startsWith("[")) {
+                // File starts with a '[' character, so it's probably in list format
+                String content = firstLine + "\n" + reader.lines().collect(Collectors.joining("\n"));
+                if (isValidJson(content)) {
```
Earlier i was parsing the entire json. Now i am just checking the first character. So no hanging of the wizard now.

teetangh commented on src/main/java/com/couchbase/intellij/tools/dialog/ImportDialog.java:
```diff
@@ -0,0 +1,1554 @@
+import utils.TemplateUtil;
+
+public class ImportDialog extends DialogWrapper {
+
```
Removed all the warnings. only typos remain

deniswsrosa commented:
- Rename "Default Scope and Collection" to "Into the Default Scope and Collection"
- Rename "Target Collection" to "Into a specified Collection"
- On the help, the scope and collection, there is no indication of what "%%" means. Please explain that to the user 
<img width="677" alt="Screenshot 2023-09-04 at 6 01 53 PM" src="https://github.com/couchbaselabs/couchbase_jetbrains_plugin/assets/216006/33ea9006-4dac-4781-ae7e-3f7f881821ac">

- Remove the option to create a collection when the scope has zero collections

Everything else seems to be fine.

deniswsrosa reviewed (changes requested):
Check my comments
